File: oauth2-extended-authorization-server/authorization-server/src/main/java/com/onlyteo/sandbox/config/WebSecurityConfig.java
package com.onlyteo.sandbox.config;

import com.onlyteo.sandbox.mapper.AppSecurityPropertiesMapper;
import com.onlyteo.sandbox.properties.AppSecurityProperties;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.annotation.Order;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.crypto.factory.PasswordEncoderFactories;
import org.springframework.security.crypto.password.DelegatingPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.provisioning.JdbcUserDetailsManager;
import org.springframework.security.provisioning.UserDetailsManager;
import org.springframework.security.web.SecurityFilterChain;

@EnableConfigurationProperties(AppSecurityProperties.class)
@EnableWebSecurity
@Configuration(proxyBeanMethods = false)
public class WebSecurityConfig {

    /**
     * The {@link SecurityFilterChain} defines how the application should be protected using Spring Security.
     * By using the DSL of the {@link HttpSecurity} configuration builder it is possible to customize the security
     * setup.
     * Here the application is configured as with Form Login. This enables the OAuth2 Authorization Code Grant for clients.
     * See the JavaDoc of the {@link HttpSecurity#formLogin} method for more details.
     *
     * @param http               - HTTP security configuration builder.
     * @param securityProperties - Custom security properties.
     * @return The {@link SecurityFilterChain} bean.
     * @throws Exception -
     */
    @Bean
    @Order(2)
    public SecurityFilterChain webSecurityFilterChain(final HttpSecurity http,
                                                      final AppSecurityProperties securityProperties) throws Exception {
        return http
                .authorizeHttpRequests(config -> config
                        .requestMatchers(securityProperties.getWhitelistedPathsAsArray()).permitAll()
                        .anyRequest().authenticated()
                )
                .formLogin(config -> config
                        .loginPage("/login").permitAll()
                )
                .build();
    }

    /**
     * The default behaviour of Spring Security is to use an in-memory {@link UserDetailsManager}. This changes
     * to use a JDBC variant so that {@link UserDetails} instances are stored in a database.
     *
     * @param jdbcTemplate       - The default {@link JdbcTemplate} bean.
     * @param passwordEncoder    - The {@link PasswordEncoder} bean from below.
     * @param securityProperties - Custom security properties.
     * @return The {@link JdbcUserDetailsManager} bean.
     */
    @Bean
    public UserDetailsManager userDetailsManager(final JdbcTemplate jdbcTemplate,
                                                 final PasswordEncoder passwordEncoder,
                                                 final AppSecurityProperties securityProperties) {
        final var propertiesMapper = new AppSecurityPropertiesMapper(passwordEncoder, securityProperties);
        final var userDetails = propertiesMapper.asUserDetails();
        final var userDetailsManager = new JdbcUserDetailsManager();
        userDetailsManager.setJdbcTemplate(jdbcTemplate);
        userDetails.stream()
                .filter(user -> !userDetailsManager.userExists(user.getUsername()))
                .forEach(userDetailsManager::createUser);
        return userDetailsManager;
    }

    /**
     * A {@link PasswordEncoder} bean.
     *
     * @return The {@link DelegatingPasswordEncoder} bean.
     */
    @Bean
    public PasswordEncoder passwordEncoder() {
        return PasswordEncoderFactories.createDelegatingPasswordEncoder();
    }
}

File: spring-boot-oauth2-authorization-server-extended/spring-boot-oauth2-authorization-server-extended-server/src/main/kotlin/com/onlyteo/sandbox/config/WebSecurityConfig.kt
package com.onlyteo.sandbox.config

import com.onlyteo.sandbox.mapper.AppSecurityPropertiesMapper
import com.onlyteo.sandbox.properties.AppSecurityProperties
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.annotation.Order
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configurers.FormLoginConfigurer
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.crypto.factory.PasswordEncoderFactories
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.provisioning.JdbcUserDetailsManager
import org.springframework.security.provisioning.UserDetailsManager
import org.springframework.security.web.SecurityFilterChain

@EnableConfigurationProperties(AppSecurityProperties::class)
@EnableWebSecurity
@Configuration(proxyBeanMethods = false)
class WebSecurityConfig {

    /**
     * The [SecurityFilterChain] defines how the application should be protected using Spring Security.
     * By using the DSL of the [HttpSecurity] configuration builder it is possible to customize the security
     * setup.
     * Here the application is configured as with Form Login. This enables the OAuth2 Authorization Code Grant for clients.
     * See the JavaDoc of the [HttpSecurity.formLogin] method for more details.
     *
     * @param http               - HTTP security configuration builder.
     * @param securityProperties - Custom security properties.
     * @return The [SecurityFilterChain] bean.
     * @throws Exception -
     */
    @Bean
    @Order(2)
    @Throws(Exception::class)
    fun webSecurityFilterChain(
        http: HttpSecurity,
        securityProperties: AppSecurityProperties
    ): SecurityFilterChain {
        return http
            .authorizeHttpRequests { config ->
                config
                    .requestMatchers(*securityProperties.whitelistedPaths.toTypedArray()).permitAll()
                    .anyRequest().authenticated()
            }
            .formLogin { config: FormLoginConfigurer<HttpSecurity?> ->
                config
                    .loginPage("/login").permitAll()
            }
            .build()
    }

    /**
     * The default behaviour of Spring Security is to use an in-memory [UserDetailsManager]. This changes
     * to use a JDBC variant so that [UserDetails] instances are stored in a database.
     *
     * @param jdbcTemplate       - The default [JdbcTemplate] bean.
     * @param passwordEncoder    - The [PasswordEncoder] bean from below.
     * @param securityProperties - Custom security properties.
     * @return The [JdbcUserDetailsManager] bean.
     */
    @Bean
    fun userDetailsManager(
        jdbcTemplate: JdbcTemplate,
        passwordEncoder: PasswordEncoder,
        securityProperties: AppSecurityProperties
    ): UserDetailsManager {
        val propertiesMapper = AppSecurityPropertiesMapper(passwordEncoder, securityProperties)
        val userDetails = propertiesMapper.asUserDetails()
        val userDetailsManager = JdbcUserDetailsManager()
        userDetailsManager.jdbcTemplate = jdbcTemplate
        userDetails.stream()
            .filter { user: UserDetails -> !userDetailsManager.userExists(user.username) }
            .forEach { user: UserDetails? -> userDetailsManager.createUser(user) }
        return userDetailsManager
    }

    /**
     * A [PasswordEncoder] bean.
     *
     * @return The [PasswordEncoder] bean.
     */
    @Bean
    fun passwordEncoder(): PasswordEncoder {
        return PasswordEncoderFactories.createDelegatingPasswordEncoder()
    }
}